_P177_CM1107.ino

```diff
@@ -0,0 +1,235 @@
+  checksum += (uint8_t)co2;
+  log += co2;
+  checksum += Wire.read(); // expected 100
+  checksum += Wire.read(); // CS
+  log += F(", checksum: ");
+  if (checksum)
+  {
+    log += F("ERR");
+    log += checksum;
+  } else
+  {
+    log += F("OK");
+  }
+  addLog(LOG_LEVEL_INFO,log);
+  return log;
+}
+
+uint16_t plugin_177_getCO2()
+  // Retrieve CO2 data in ppm
+{
+  I2C_write8(CM1107_ADDR, CM1107_CMD_CO2);
+  uint16_t co2;
+  uint8_t status;
+  uint8_t checksum;
+
+  Wire.requestFrom(CM1107_ADDR, (byte)5);
+  checksum = Wire.read(); // CM1107_CMD_CO2
+  co2 = Wire.read(); // high byte ppm
+  checksum += (uint8_t)co2;
+  co2 = (co2 << 8) | Wire.read(); // low byte ppm
+  checksum += (uint8_t)co2;
+  status = Wire.read(); // status
+  checksum += status;
+  checksum += Wire.read(); //CS
+
+  String log = F("CM1107: CO2 ppm:");
+  log += co2;
+  log += F(", status: ");
+  log += status;
+  log += F(", checksum: ");
+  if (checksum)
+  {
+    log += F("ERR");
+    log += checksum;
+  } else
+  {
+    log += F("OK");
+  }
+  addLog(LOG_LEVEL_INFO,log);
+  return co2;
+}
+
+
+boolean Plugin_177(byte function, struct EventStruct *event, String& string)
+{
+  boolean success = false;
+
+  switch (function)
+  {
+    case PLUGIN_DEVICE_ADD:
+      {
+        Device[++deviceCount].Number = PLUGIN_ID_177;
+        Device[deviceCount].Type = DEVICE_TYPE_I2C;
+        Device[deviceCount].VType = SENSOR_TYPE_SINGLE;
+        Device[deviceCount].Ports = 0;
+        Device[deviceCount].PullUpOption = false;
+        Device[deviceCount].InverseLogicOption = false;
+        Device[deviceCount].FormulaOption = true;
+        Device[deviceCount].ValueCount = 1;
+        Device[deviceCount].SendDataOption = true;
+        Device[deviceCount].TimerOption = true;
+        Device[deviceCount].GlobalSyncOption = true;
+        break;
+      }
+
+    case PLUGIN_GET_DEVICENAME:
+      {
+        string = F(PLUGIN_NAME_177);
+        break;
+      }
+
+    case PLUGIN_GET_DEVICEVALUENAMES:
+      {
+        strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_177));
+        break;
+      }
+
+    case PLUGIN_WEBFORM_LOAD:
+      {
+        addFormNumericBox(F("ABC period"), F("plugin_177_CM1107_abc"), Settings.TaskDevicePluginConfig[event->TaskIndex][0]);
+        addUnit(F("days, 0=off"));
+        success = true;
+        break;
+      }
+
+    case PLUGIN_WEBFORM_SAVE:
+      {
+        unsigned char abc = getFormItemInt(F("plugin_177_CM1107_abc"));
+        if (abc > 15) abc = 15;
+        Settings.TaskDevicePluginConfig[event->TaskIndex][0] = abc;
+        success = true;
+        break;
+      }
+    case PLUGIN_INIT:
+      {
+        plugin_177_begin();
+        success = plugin_177_setABC((uint8_t) Settings.TaskDevicePluginConfig[event->TaskIndex][0]);
+        I2C_write8(CM1107_ADDR, CM1107_CMD_CO2);
+        break;
+      }
+    case PLUGIN_READ:
+      {
+        plugin_177_begin();
+        co2 = plugin_177_getCO2();
```

Comment:
Where is this variable `co2` declared?
If this code is working, then I'm really curious to see where it is being declared and what value is being messed with by this code.

Reply:
Fixed, missed declaration. Code worked because of same name global variable from my different plugin.